### PR TITLE
feat(examples): TorchRL - MAML integration

### PR DESCRIPTION
**Pull request by vmoens**
Brand-new version of the old PR

One can test this by running
```python
python maml_torchrl.py --parallel --seed 1
```

I have added the `--parallel` flag, to allow users to have more than one environment used. 

It can run on cuda if devices are present.

FYI: This should run on latest `main` in torchrl

### CODE REVIEW

**Review by Benjamin-eecs — CHANGES_REQUESTED** [review]
@vmoens Hey Vincent, I've adapted the maml_torchrl script using lastest torchrl and corrected my own MAML-RL implementation, but the results are still not aligned. Mind taking a look why that happened?

**Comment by vmoens** [comment]
> @vmoens Hey Vincent, I've adapted the maml_torchrl script using lastest torchrl and corrected my own MAML-RL implementation, but the results are still not aligned. Mind taking a look why that happened?

Thanks for this work!
I see that MetaSGD has a lr=0.3 for torchrl but 0.1 for the other version, is that intended?

**Comment by Benjamin-eecs** [comment]
> > @vmoens Hey Vincent, I've adapted the maml_torchrl script using lastest torchrl and corrected my own MAML-RL implementation, but the results are still not aligned. Mind taking a look why that happened?
> 
> Thanks for this work! I see that MetaSGD has a lr=0.3 for torchrl but 0.1 for the other version, is that intended?

Oops, I forgot to run the torchrl with inner lr=0.1. I think we should lower the inner learning rate all to lr=0.1. The performance figure now is showing the results with lr=0.1/

**Comment by vmoens** [comment]
> > > @vmoens Hey Vincent, I've adapted the maml_torchrl script using lastest torchrl and corrected my own MAML-RL implementation, but the results are still not aligned. Mind taking a look why that happened?
> > 
> > 
> > Thanks for this work! I see that MetaSGD has a lr=0.3 for torchrl but 0.1 for the other version, is that intended?
> 
> Oops, I forgot to run the torchrl version of MAML with inner lr=0.1. I think we should lower the inner learning rate all to lr=0.1. The performance figure now is showing the results with lr=0.1.

Sorry just to make sure I understood well: the PR indicates that lr=0.3 but the test was run with lr=0.1, right?

**Comment by Benjamin-eecs** [comment]
> > > > @vmoens Hey Vincent, I've adapted the maml_torchrl script using lastest torchrl and corrected my own MAML-RL implementation, but the results are still not aligned. Mind taking a look why that happened?
> > > 
> > > 
> > > Thanks for this work! I see that MetaSGD has a lr=0.3 for torchrl but 0.1 for the other version, is that intended?
> > 
> > 
> > Oops, I forgot to run the torchrl version of MAML with inner lr=0.1. I think we should lower the inner learning rate all to lr=0.1. The performance figure now is showing the results with lr=0.1.
> 
> Sorry just to make sure I understood well: the PR indicates that lr=0.3 but the test was run with lr=0.1, right?

Oh, I was mentioning the performance figure with my implementation. The torchrl performance figure in this PR was runned with lr=0.3. We may have to change the lr to 0.3 and run it again.

**Comment by Benjamin-eecs** [comment]
> > @vmoens Hey Vincent, I've adapted the maml_torchrl script using lastest torchrl and corrected my own MAML-RL implementation, but the results are still not aligned. Mind taking a look why that happened?
> 
> Thanks for this work! I see that MetaSGD has a lr=0.3 for torchrl but 0.1 for the other version, is that intended?

I've already updated the results with lr=0.1, still not learning with torchrl implementation, mind taking a closer look at it?

**Comment by vmoens** [comment]
> I've already updated the results with lr=0.1, still not learning with torchrl implementation, mind taking a closer look at it?

Of course.
Has TorchOpt dropped cpu support? I used to be able to run it on my mcbook but it seems that now nvcc is needed.
It's harder to debug on remote clusters...

**Comment by XuehaiPan** [comment]
> Has TorchOpt dropped cpu support? I used to be able to run it on my mcbook but it seems that now nvcc is needed.

@vmoens If you are installing `TorchOpt` from source, you will need a `nvcc` to compile the `.cu` files. We provide a conda environment with fully-capable build toolchains.

```bash
git clone https://github.com/metaopt/TorchOpt.git
cd TorchOpt

# This branch bumped torch == 1.12.1 and functorch == 0.2.1 (which were released yesterday)
git checkout release/0.4.3

CONDA_OVERRIDE_CUDA=11.7 conda env create --file conda-recipe.yaml

conda activate torchopt
pip3 install --no-build-isolation --editable .
```

Note that you do not need a physical GPU installed on the machine to compile the source code.

------

Alternative, if you don't want to create a new conda environment or install `TorchOpt` from PyPI, you will need to build a new wheel (there was a bug in the wheels on PyPI, see #46 and #45). You can build a new wheel using `cibuildwheel` and `docker`:

```bash
git clone https://github.com/metaopt/TorchOpt.git
cd TorchOpt

# This branch bumped torch == 1.12.1 and functorch == 0.2.1 (which were released yesterday)
git checkout release/0.4.3

python3 -m pip install --upgrade cibuildwheel  # you also need `docker` installed
CIBW_BUILD="cp39*manylinux*" python3 -m cibuildwheel --plat=linux --config-file=pyproject.toml
```

You can change `cp39` in the variable `CIBW_BUILD` to your specific Python version (e.g. `"cp38*manylinux*" ` for Python 3.8). A new wheel will list in the `wheelhouse/` folder. Then install it by:

```bash
python3 -m pip install wheelhouse/torchopt-<tag>.whl
```

File: examples/MAML-RL/maml_torchrl.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    # Env
+    lambda_env = lambda: GymEnv(
+        'TabularMDP-v0',
+        num_states=STATE_DIM,
+        num_actions=ACTION_DIM,
+        max_episode_steps=TRAJ_LEN,
+        seed=args.seed,
+    )
+    if args.parallel:
+        env = ParallelEnv(
```
Comment (vmoens) [inline examples/MAML-RL/maml_torchrl.py:127]: I guess we should seed the env too?

File: examples/MAML-RL/maml_torchrl.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    # init training
+    torch.manual_seed(args.seed)
+    torch.cuda.manual_seed_all(args.seed)
+    # Env
+    lambda_env = lambda: GymEnv(
+        'TabularMDP-v0',
+        num_states=STATE_DIM,
+        num_actions=ACTION_DIM,
+        max_episode_steps=TRAJ_LEN,
+        seed=args.seed,
```
Comment (vmoens) [inline examples/MAML-RL/maml_torchrl.py:124]: Let's avoid that: all envs will have the same seed! (that's my n1 hypothesis regarding the diff in perf between the 2 implementations)

**Review by vmoens — COMMENTED** [review]
I'm testing this with a different seed for each env, bear with me, i'll come up with some results tomorrow!

File: examples/MAML-RL/maml_torchrl.py (hunk 1 above)
Comment (Benjamin-eecs) [reply examples/MAML-RL/maml_torchrl.py:127]: Sure.

File: examples/MAML-RL/maml_torchrl.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+    tasks = dummy_env.sample_tasks(num_tasks=task_num)
+
+    policy_state_dict = torchopt.extract_state_dict(actor_critic)
+    optim_state_dict = torchopt.extract_state_dict(inner_opt)
+    for idx in range(task_num):
+        env.reset_task(tasks[idx])
+        for _ in range(inner_iters):
+            with set_exploration_mode('random'), torch.no_grad(), timeit('rollout_eval'):
+                pre_traj_td = env.rollout(policy=policy, max_steps=TRAJ_LEN).to(device)
```
[the remark this answers is not in the document]
Comment (Benjamin-eecs) [reply examples/MAML-RL/maml_torchrl.py]: I am not really sure what you meant, but I think giving the done state the current reward without summation of any other discounting reward from other time step is the correct implementation.

**Review by XuehaiPan — DISMISSED** [review]

**Comment by vmoens** [comment]
Yes it is on train, just needs better sync on test wrt the seed setting. (At least on my machine, let me update the plot).

Results aren't quite different tbh, and we're still far from significant learning improvement in the original version (the error bars are hugely overlapping)
I'm happy to do some more hyperparameter optimisation to match the other results better.

**Review by XuehaiPan — APPROVED** [review]

**Review by XuehaiPan — APPROVED** [review]

**Review by Benjamin-eecs — APPROVED** [review]
